### PR TITLE
Feature/24 show forks optimization

### PR DESCRIPTION
**Pull request by TE4-johan-corneliusson**
API-Calls happen pararell instead of sequential #24 
Also added buffer bar, kinda in the wrong branch, i'm sorry

### CODE REVIEW

**Comment by TE4-johan-corneliusson** [comment]
Also added some error checking stuff

**Comment by TE4-johan-corneliusson** [comment]
Komm precis på att node_modules är med i .gitignore, så detta kommer snea allt eftersom jag ändrat filsökvägar så att buffer bar skulle fungera.

File: assets/css/style.css [hunk 1]
```diff
@@ -1,3 +1,5 @@
+@import "/node_modules/@material/linear-progress/dist/mdc.linear-progress.css";
```
Comment (itggot-linus-styren) [inline assets/css/style.css]: What do you think of using a relative path instead: `@import "@material/linear-progress/mdc-linear-progress";`?

File: assets/js/index.js [hunk 2]
```diff
@@ -1,35 +1,50 @@
+import { MDCLinearProgress } from '/node_modules/@material/linear-progress/index.js';
```
Comment (itggot-linus-styren) [inline assets/js/index.js]: Could you explain why `/node_modules` prefix is required? My suggestion is to use `import { MDCLinearProgress } from '@material/linear-progress';`?

File: assets/js/index.js [hunk 3]
```diff
@@ -1,35 +1,50 @@
...
     const response = await fetch(`https://api.github.com/users/${userName}/repos?client_id=01b71e233f58e4f88a33?client_secret=b1dd13a0dd21fb36d4a496d094c9b8ade6ea9e53`)
     const result = await response.json();
     return result
 }
 
 async function getForks(userName, repo) {
-    const responseFork = await fetch(`https://api.github.com/repos/${userName}/${repo}/forks?client_id=01b71e233f58e4f88a33?client_secret=b1dd13a0dd21fb36d4a496d094c9b8ade6ea9e53`)
-    const resultFork = await responseFork.json();
-    return resultFork
+    const responseForks = await fetch(`https://api.github.com/repos/${userName}/${repo}/forks?client_id=01b71e233f58e4f88a33?client_secret=b1dd13a0dd21fb36d4a496d094c9b8ade6ea9e53`)
```
Comment (itggot-linus-styren) [inline assets/js/index.js]: If possible, running lint --fix should normalise all uses of semicolons.

File: assets/js/index.js [hunk 4]
```diff
@@ -1,35 +1,50 @@
...
+    const resultForks = await responseForks.json();
+    const manifest = await getManifestFile(userName, repo)
+    if (manifest.error) return null
+
+    //add code to each corrosponding fork with pararell api calls
+    const promises = resultForks.map(fork => getStudentsCode(fork.owner.login, repo, manifest.filePath))
+    const codes = await Promise.all(promises)
+    for (let i = 0; i < resultForks.length; i++) {
```
Comment (itggot-linus-styren) [inline assets/js/index.js]: Does the `codes` array retain order? Perhaps it should be sorted alphabetically.

File: assets/js/index.js [hunk 5]
```diff
@@ -1,35 +1,50 @@
...
+        resultForks[i].code = codes[i]
+    }
+    return resultForks
 }
 
 
 async function getStudentsCode(userName, repo, path){
     const response = await fetch(`https://api.github.com/repos/${userName}/${repo}/contents/${path}?client_id=01b71e233f58e4f88a33?client_secret=b1dd13a0dd21fb36d4a496d094c9b8ade6ea9e53`)
+    if (response.ok == false) return `'Could not find code at "${path}" in repo "${repo}" owned by user "${userName}"'`
```
Comment (itggot-linus-styren) [inline assets/js/index.js]: Why the extra single quotes?

File: assets/js/index.js [hunk 6]
```diff
@@ -1,35 +1,50 @@
...
 
 async function getManifestFile(userName, repo) {
+
     const response = await fetch(`https://api.github.com/repos/${userName}/${repo}/contents/.manifest.json?client_id=01b71e233f58e4f88a33?client_secret=b1dd13a0dd21fb36d4a496d094c9b8ade6ea9e53`)
+    if (response.ok == false) {
+        let error = { 
+            error: true,
+            text: `No ".manifest.json" found in repo "${repo}" owned by user "${userName}"`
+        }
+        showError(error)
```
Comment (itggot-linus-styren) [inline assets/js/index.js]: Good error handling, maybe this should be used on the above error message as well?

File: assets/css/style.css (hunk 1 above)
Comment (TE4-johan-corneliusson) [reply assets/css/style.css]: Jo det är så documentationen säger att man ska göra, men det funkar inte.

File: assets/js/index.js (hunk 2 above)
Comment (TE4-johan-corneliusson) [reply assets/js/index.js]: fungerar inte heller. kan jag förklara varför: nej!

File: assets/js/index.js (hunk 4 above)
Comment (TE4-johan-corneliusson) [reply assets/js/index.js]: Promise.all() does indeed retain order

File: assets/js/index.js (hunk 5 above)
Comment (TE4-johan-corneliusson) [reply assets/js/index.js]: So that this code won't be syntax highlighted as it's displayed in a code field.

**Pull request merged**